pkg/controller/pxc/vs.go

```diff
@@ -0,0 +1,123 @@
+package pxc
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+func (vs VersionServiceMock) Apply(desiredVersion string, currentVersion string) (DepVersion, error) {
+	client := http.Client{
+		Timeout: 5 * time.Second,
+	}
+	url := fmt.Sprintf("%s/api/versions/v1/pxc/%s/%s?databaseVersion=%s", vs.URL, vs.OpVersion, desiredVersion, currentVersion)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return DepVersion{}, err
+	}
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return DepVersion{}, err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return DepVersion{}, fmt.Errorf("received bad status code %s", resp.Status)
+	}
+
+	r := VersionResponse{}
+	err = json.NewDecoder(resp.Body).Decode(&r)
+	if err != nil {
+		return DepVersion{}, fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	if len(r.Versions) == 0 {
+		return DepVersion{}, fmt.Errorf("empty versions response")
+	}
+	pxcVersion, err := getVersion(r.Versions[0].Matrix.PXC)
+	if err != nil {
+		return DepVersion{}, err
+	}
+	backupVersion, err := getVersion(r.Versions[0].Matrix.Backup)
+	if err != nil {
+		return DepVersion{}, err
+	}
+	pmmVersion, err := getVersion(r.Versions[0].Matrix.PMM)
+	if err != nil {
+		return DepVersion{}, err
+	}
+	proxySqlVersion, err := getVersion(r.Versions[0].Matrix.ProxySQL)
+	if err != nil {
+		return DepVersion{}, err
+	}
+
+	return DepVersion{
+		PXCImage:        r.Versions[0].Matrix.PXC[pxcVersion].ImagePath,
+		PXCVersion:      pxcVersion,
+		BackupImage:     r.Versions[0].Matrix.Backup[backupVersion].ImagePath,
+		BackupVersion:   backupVersion,
+		ProxySqlImage:   r.Versions[0].Matrix.ProxySQL[proxySqlVersion].ImagePath,
+		ProxySqlVersion: proxySqlVersion,
+		PMMImage:        r.Versions[0].Matrix.PMM[pmmVersion].ImagePath,
+		PMMVersion:      pmmVersion,
+	}, nil
+}
+
+func getVersion(versions map[string]Version) (string, error) {
+	if len(versions) == 0 {
+		return "", fmt.Errorf("failed to get version from empty map")
+	}
+
+	for k := range versions {
+		return k, nil
+	}
+	return "", nil
+}
```

Comment:
1. Why do we return a random version?
2. Shouldn't be "status" and "critical" fields taken into account?

Reply:
only the one version must be returned, we should choose the first in the list

Reply:
Exactly, we assume that in response we have just one version

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/80c40ef050a99f663325d529f15826d3694b5f64